eng/scripts/test-matrix/test-matrix-functions.ps1

```diff
@@ -0,0 +1,317 @@
+Set-StrictMode -Version "4.0"
+
+function CreateDisplayName([string]$parameter, [hashtable]$displayNames) {
+    $name = $parameter
+
+    if ($displayNames[$parameter]) {
+        $name = $displayNames[$parameter]
+    }
+
+    return $name -replace "[\-\._]", ""
+}
+
+function GenerateMatrix([HashTable]$config, [string]$selectFromMatrixType) {
+    if ($selectFromMatrixType -eq "sparse") {
+        [Array]$dimensions = GetMatrixDimensions $config.matrix
+        $size = $dimensions[0]
+        [Array]$matrix = GenerateSparseMatrix $config.matrix $config.displayNames $size
+    } elseif ($selectFromMatrixType -eq "all") {
+        [Array]$matrix, $_ = GenerateFullMatrix $config.matrix $config.displayNames
+    } else {
+        throw "Matrix generator not implemented for selectFromMatrix: $($platform.selectFromMatrix)"
+    }
+
+    if ($config["exclude"]) {
+        [Array]$matrix = ProcessExcludes $matrix $config.exclude
+    }
+    if ($config["include"]) {
+        [Array]$matrix = ProcessIncludes $matrix $config.include $config.displayNames
+    }
+
+    return $matrix | Sort-Object -Property name
+}
+
+function ProcessExcludes([Array]$matrix, [Array]$excludes) {
+    $deleteKey = "%DELETE%"
+    $exclusionMatrix = @()
+
+    foreach ($exclusion in $excludes) {
+        $converted = ConvertToMatrixArrayFormat $exclusion
+        $full, $_ = GenerateFullMatrix $converted
+        $exclusionMatrix += $full
+    }
+
+    foreach ($element in $matrix) {
+        foreach ($exclusion in $exclusionMatrix) {
+            $match = MatrixElementMatch $element.parameters $exclusion.parameters
+            if ($match) {
+                $element.parameters[$deleteKey] = $true
+            }
+        }
+    }
+
+    return $matrix | Where-Object { !$_.parameters.Contains($deleteKey) }
+}
+
+function ProcessIncludes([Array]$matrix, [Array]$includes, [HashTable]$displayNames) {
+    foreach ($inclusion in $includes) {
+        $converted = ConvertToMatrixArrayFormat $inclusion
+        $full, $_ = GenerateFullMatrix $converted $displayNames
+        $matrix += $full
+    }
+
+    return $matrix
+}
+
+function MatrixElementMatch([HashTable]$source, [HashTable]$target) {
+    if ($target.Count -eq 0) {
```

Comment:
Should this also include a check to ensure the Count of the source and target are equal?

Reply:
The counts do not need to be equal intentionally. The exclude only needs to match all of its specified fields, even if the matrix entry has more. For example, if we wanted to exclude all windows tests on .net core, we wouldn't need to worry about specifying all other values. I have a note about this [here](https://github.com/Azure/azure-sdk-for-net/blob/463e3d15352ad110eca82add53ecef400e9762ec/eng/scripts/test-matrix/README.md#exclude)